File: src/main/java/nz/co/searchwellington/repositories/SuggestedFeeditemsService.java
package nz.co.searchwellington.repositories;

import java.util.List;

import nz.co.searchwellington.feeds.FeedItemLocalCopyDecorator;
import nz.co.searchwellington.feeds.reading.WhakaokoFeedItemMapper;
import nz.co.searchwellington.feeds.reading.WhakaoroClientFactory;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.FeedAcceptancePolicy;
import nz.co.searchwellington.model.frontend.FeedNewsitemForAcceptance;
import nz.co.searchwellington.model.frontend.FrontendFeedNewsitem;
import nz.co.searchwellington.model.frontend.FrontendNewsitem;

import org.apache.log4j.Logger;
import org.elasticsearch.common.collect.Lists;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import uk.co.eelpieconsulting.common.http.HttpBadRequestException;
import uk.co.eelpieconsulting.common.http.HttpFetchException;
import uk.co.eelpieconsulting.common.http.HttpForbiddenException;
import uk.co.eelpieconsulting.common.http.HttpNotFoundException;
import uk.co.eelpieconsulting.whakaoro.client.exceptions.ParsingException;
import uk.co.eelpieconsulting.whakaoro.client.model.FeedItem;

@Component
public class SuggestedFeeditemsService {
	
	private static final Logger log = Logger.getLogger(SuggestedFeeditemsService.class);
	
	private final FeedItemLocalCopyDecorator feedItemLocalCopyDecorator;
	private final WhakaoroClientFactory whakaoroClientFactory;
	private final WhakaokoFeedItemMapper whakaokoFeedItemMapper;
	private final HibernateResourceDAO resourceDAO;
	
	@Autowired
	public SuggestedFeeditemsService(WhakaoroClientFactory whakaoroClientFactory, 
			WhakaokoFeedItemMapper whakaokoFeedItemMapper,
			FeedItemLocalCopyDecorator feedItemLocalCopyDecorator,
			HibernateResourceDAO resourceDAO) {
		this.whakaoroClientFactory = whakaoroClientFactory;
		this.whakaokoFeedItemMapper = whakaokoFeedItemMapper;		
		this.feedItemLocalCopyDecorator = feedItemLocalCopyDecorator;
		this.resourceDAO = resourceDAO;
	}
	
	public List<FrontendNewsitem> getSuggestionFeednewsitems(int maxItems) {
		try {
			List<FrontendFeedNewsitem> channelFeedItemsForNotIgnoredFeeds = Lists.newArrayList();
			for (FeedItem feedItem : whakaoroClientFactory.getChannelFeedItems()) {
				final Feed feed = resourceDAO.loadFeedByWhakaoroId(feedItem.getSubscriptionId());
				if (feed == null) {
					log.info("Ignoring feed item with unknown whakaoro id: " + feedItem.getSubscriptionId());
					continue;
				}
				if (feed.getAcceptancePolicy().equals(FeedAcceptancePolicy.IGNORE)) {
					continue;
				}
				channelFeedItemsForNotIgnoredFeeds.add(whakaokoFeedItemMapper.mapWhakaokoFeeditem(feed, feedItem));
			}
			
			List<FeedNewsitemForAcceptance> addSupressionAndLocalCopyInformation = feedItemLocalCopyDecorator.addSupressionAndLocalCopyInformation(channelFeedItemsForNotIgnoredFeeds);
			
			List<FrontendNewsitem> suggestions = Lists.newArrayList();
			for (FeedNewsitemForAcceptance feedNewsitemForAcceptance : addSupressionAndLocalCopyInformation) {
				if (feedNewsitemForAcceptance.getAcceptanceState().getLocalCopy() != null) {
					continue;
				}
				suggestions.add(feedNewsitemForAcceptance.getFeednewsitem());
			}
			return suggestions;
			
		} catch (HttpNotFoundException e) {
			log.error(e);
		} catch (HttpBadRequestException e) {
			log.error(e);
		} catch (HttpForbiddenException e) {
			log.error(e);
		} catch (ParsingException e) {
			log.error(e);
		} catch (HttpFetchException e) {
			log.error(e);
		}
		
		return Lists.newArrayList();
	}
	
}

File: src/main/scala/nz/co/searchwellington/repositories/SuggestedFeeditemsService.scala
package nz.co.searchwellington.repositories

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.feeds.{FeedItemLocalCopyDecorator, FeeditemToNewsitemService, RssfeedNewsitemService}
import nz.co.searchwellington.model.{Newsitem, User}
import nz.co.searchwellington.model.frontend.FrontendResource
import nz.co.searchwellington.model.mappers.FrontendResourceMapper
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.apache.log4j.Logger
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.concurrent.{ExecutionContext, Future}

@Component class SuggestedFeeditemsService @Autowired()(rssfeedNewsitemService: RssfeedNewsitemService,
                                                        feedItemLocalCopyDecorator: FeedItemLocalCopyDecorator,
                                                        feeditemToNewsitemService: FeeditemToNewsitemService,
                                                        frontendResourceMapper: FrontendResourceMapper,
                                                        mongoRepository: MongoRepository, suppressionDAO: SuppressionDAO) extends
  ReasonableWaits {

  private val log = Logger.getLogger(classOf[SuggestedFeeditemsService])

  private val MaximumChannelPagesToScan = 5

  def getSuggestionFeednewsitems(maxItems: Int, loggedInUser: Option[User])(implicit ec: ExecutionContext): Future[Seq[FrontendResource]] = {

    def canBeAccepted(feedItem: FrontendResource): Future[Boolean] = Future.successful(false)

    def filteredPage(page: Int, output: Seq[FrontendResource]): Future[Seq[FrontendResource]] = {
      log.info("Fetching filter page: " + page + "/" + output.size)
      rssfeedNewsitemService.getChannelFeedItems(page).flatMap { channelFeedItems =>
        log.info("Found " + channelFeedItems.size + " channel newsitems on page " + page)

        val channelNewsitems = channelFeedItems.map(i => feeditemToNewsitemService.makeNewsitemFromFeedItem(i._1, i._2))

        val eventuallyFiltered = channelNewsitems.map { newsitem =>
          val eventuallyLocalCopy = mongoRepository.getResourceByUrl(newsitem.page)
          val eventuallyIsSuppressed = suppressionDAO.isSupressed(newsitem.page)
          for {
            localCopy <- eventuallyLocalCopy
            isSuppressed <- eventuallyIsSuppressed
          } yield {
            if (localCopy.nonEmpty || isSuppressed) {
              None
            } else {
              Some(newsitem)
            }
          }
        }

        Future.sequence(eventuallyFiltered).flatMap { filterResults =>
          val eventualFrontendChannelResources: Future[Seq[FrontendResource]] = Future.sequence {
            val filtered = filterResults.flatten
            log.info("After filtering out those with local copies: " + filtered.size)
            filtered.map { r =>
              frontendResourceMapper.mapFrontendResource(r, r.geocode)
            }
          }

          eventualFrontendChannelResources.flatMap { rs =>
            feedItemLocalCopyDecorator.withFeedItemSpecificActions(rs, loggedInUser).flatMap { suggestions =>
              log.info("Adding " + suggestions.size + " to " + output.size)
              val result = output ++ suggestions
              if (result.size >= maxItems || channelFeedItems.isEmpty || page == MaximumChannelPagesToScan) {
                Future.successful(result.take(maxItems))
              } else {
                filteredPage(page + 1, result)
              }
            }
          }
        }
      }
    }

    filteredPage(1, Seq.empty)
  }

}
